File: src/Wc3HandleLua/Handle/Image/Arc.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local Color = Wc3Utils.Color or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type ImagePixelClass
local Pixel = require('Handle.Image.Pixel') or error('')
---@type HandleSettings
local Settings = require('Settings') or error('')

--=======
-- Class
--=======

local ImageArc = Class.new('ImageArc')
---@class ImageArc
local public = ImageArc.public
---@class ImageArcClass
local static = ImageArc.static
---@type ImageArcClass
local override = ImageArc.override
local private = {}

--=========
-- Static
--=========

---@param max_p_count number
---@param p_size number | nil
---@param step number | nil
---@param child ImageArc | nil
---@return ImageArc
function override.new(max_p_count, p_size, step, child)
    isTypeErr(max_p_count, 'number', 'max_p_count')
    if p_size then isTypeErr(p_size, 'number', 'p_size') end
    if step then isTypeErr(step, 'number', 'step') end
    if child then isTypeErr(child, ImageArc, 'child') end

    local instance = child or Class.allocate(ImageArc)
    private.newData(instance, max_p_count, p_size, step)

    return instance
end

--========
-- Public
--========

---@param x0 number
---@param y0 number
---@param range number
---@param start_angle number
---@param end_angle number
function public:setPosPolar(x0, y0, range, start_angle, end_angle)
    isTypeErr(self, ImageArc, 'self')
    isTypeErr(x0, 'number', 'x0')
    isTypeErr(y0, 'number', 'y0')
    isTypeErr(range, 'number', 'range')
    isTypeErr(start_angle, 'number', 'start_angle')
    isTypeErr(end_angle, 'number', 'end_angle')
    local priv = private.data[self]
    
    if not priv.is_visible then
        return
    end

    local length = range * (end_angle - start_angle)
    local need = math.floor(length / priv.step) + 1
    if need > #priv.pixels then
        Log:wrn('ImageLine: not enought pixels to fill.')
    end

    local used = math.min(need, #priv.pixels)
    local a0 = start_angle
    local step_a = (end_angle - start_angle) / used
    for i = 1, used do
        local pixel = priv.pixels[i]
        pixel:setPos(x0 + range * math.cos(a0),
                     y0 + range * math.sin(a0))
        pixel:setVisible(true)

        a0 = a0 + step_a
    end

    for i = used, #priv.pixels do
        priv.pixels[i]:setVisible(false)
    end
end

---@param color Color
function public:setColor(color)
    isTypeErr(self, ImageArc, 'self')
    isTypeErr(color, Color, 'color')
    local priv = private.data[self]

    priv.color = Color.copy(color)
    for i = 1, #priv.pixels do
        priv.pixels[i]:setColor(color)
    end
end

---@return Color
function public:getColor()
    isTypeErr(self, ImageArc, 'self')
    return Color.copy(private.data[self].color)
end

---@param flag boolean
function public:renderAlways(flag)
    isTypeErr(self, ImageArc, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    local priv = private.data[self]

    priv.render_always = flag
    for i = 1, #priv.pixels do
        priv.pixels[i]:renderAlways(flag)
    end
end

---@param flag boolean
function public:setVisible(flag)
    isTypeErr(self, ImageArc, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    local priv = private.data[self]

    priv.is_visible = flag
    for i = 1, #priv.pixels do
        priv.pixels[i]:setVisible(flag)
    end
end

function public:destroy()
    isTypeErr(self, ImageArc, 'self')
    local priv = private.data[self]

    for i = 1, #priv.pixels do
        priv.pixels[i]:destroy()
    end
end

--=========
-- Private
--=========

private.data = setmetatable({}, {__mode = 'k'})

---@param self ImageArc
---@param max_p_count number
---@param p_size number
---@param step number
function private.newData(self, max_p_count, p_size, step)
    local priv = {
        color = Color.new(1, 1, 1, 1),
        render_always = true,
        is_visible = false,

        step = step or Settings.ImagePixelDefaultSize / 2,
        pixels = {}
    }
    private.data[self] = priv

    for i = 1, max_p_count do
        local pixel = Pixel.new(p_size)
        pixel:setColor(priv.color)
        pixel:renderAlways(priv.render_always)
        pixel:setVisible(priv.is_visible)

        table.insert(priv.pixels, pixel)
    end
end

return static

File: src/Handle/Image/Arc.ts
import { Color, Log } from '../../Utils'
import { Pixel } from "./Pixel";

export class Arc {
    constructor(max_pixels: number,
                pixel_size?: number,
                pixel_step?: number){

        this._size = pixel_size ? pixel_size : 16
        this._step = pixel_step ? pixel_step : 8
        this._in_use = 0
        this._pixels = []
        
        for (let i = 0; i < max_pixels; i++){
            let pixel = new Pixel(this._size)
            this._pixels.push(pixel)
            pixel.renderAlways = true
            pixel.visible = false
        }
    }

    public setPolarPos(x1: number, y1: number, range: number, a1: number, a2: number){
        this._x1 = x1
        this._y1 = y1
        this._range = range
        this._a1 = a1
        this._a2 = a2

        this.update()
    }

    public get x1(){return this._x1}
    public get y1(){return this._y1}
    public get range(){return this._range}
    public get a1(){return this._a1}
    public get a2(){return this._a2}
    
    public get color(){return new Color(this._color)}
    public set color(color: Color){
        this._color = new Color(color)
        for (let pixel of this._pixels){
            pixel.color = color
        }
    }

    public get renderAlways(){return this._render_always}
    public set renderAlways(flag: boolean){
        this._render_always = flag
        for (let pixel of this._pixels){
            pixel.renderAlways = flag
        }
    }

    public get visible(){return this._visible}
    public set visible(flag: boolean){
        this._visible = flag
        for (let i = 0; i < math.min(this._in_use, this._pixels.length); i++){
            this._pixels[i].visible = flag
        }
    }

    public destroy(){
        for (let pixel of this._pixels){
            pixel.destroy()
        }
    }

    private update(){
        let da = this._a2 - this._a1
        let arc_length = this._range * (da)

        this._in_use = math.floor(arc_length / this._step) + 1
        if (this._in_use > this._pixels.length){
            Log.wrn(Arc.toString() + 
                    ': is not enought pixels to fill arc.')
        }

        let a = this.a1
        let step_a = da / this._in_use
        let rly_used = math.min(this._in_use, this._pixels.length)

        for (let i = 0; i < rly_used; i++){
            let pixel = this._pixels[i]
            pixel.visible = this._visible
            pixel.x = this._x1 + this._range * Cos(a)
            pixel.y = this._y1 + this._range * Sin(a)
            a += step_a
        }

        for (let i = rly_used; i < this._pixels.length; i++){
            this._pixels[i].visible = false
        }
    }

    private _x1: number = 0;
    private _y1: number = 0;
    private _range: number = 0;
    private _a1: number = 0;
    private _a2: number = 0;

    private _size: number;
    private _step: number;
    private _in_use: number;
    private _pixels: Pixel[];

    private _color = new Color(1, 1, 1, 1)
    private _render_always = true
    private _visible = false
}